### PR TITLE
REF/TST: add seed to hdrboxplot, Use random order in pytest

### PR DESCRIPTION
**Pull request by bashtage**
Use random order to avoid intertest dependency

Tests show there are some problems with order affecting test success.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Closes #5110

**Comment by bashtage** [comment]
xref #4965

**Comment by josef-pkt** [comment]
Is there a way to replicate a specific random order when we need to investigate a failure?

**Comment by bashtage** [comment]
You can pass the seed argument to reproduce the same order

**Comment by josef-pkt** [comment]
Ok, those are mainly regression tests but it doesn't set the seed in random panel. 
I'm not sure why it doesn't always fails. I guess some other function sets the seed and the numpy global RandomState before this test (setup_class) is executed.

**Comment by josef-pkt** [comment]
PR for Theil test is #5149 
I will merge that when it's green

**Comment by bashtage** [comment]
I think this can be merged after I rebase and it gets green.  I added a fix that only seems to affect appveyor.

File: statsmodels/graphics/tests/test_functional.py [hunk 1]
```diff
@@ -116,36 +121,37 @@ def test_hdr_bw(close_figures):
...
+    expected_depth = [0.5, 5. / 6, 5. / 6, 0.5]
     assert_almost_equal(depth, expected_depth)
-
     ## Plot to visualize why we expect this output
-    #fig = plt.figure()
-    #ax = fig.add_subplot(111)
-    #for ii, yy in enumerate([y1, y2, y3, y4]):
+    # fig = plt.figure()
+    # ax = fig.add_subplot(111)
+    # for ii, yy in enumerate([y1, y2, y3, y4]):
```
Comment (jbrockmendel) [inline statsmodels/graphics/tests/test_functional.py]: Are these worth keeping as comments?  Or actually just defunct code?  If it's the latter then second-best (after deleting) is to keep them pep8-invalid so they don't fall through the cracks.

**Comment by bashtage** [comment]
@josef-pkt This is ready now.  The changes to HDR are all backward compat and are both good practice and helpful in testing.

File: statsmodels/graphics/functional.py [hunk 2]
```diff
@@ -392,7 +410,8 @@ def _band_quantiles(band):
 
         pool = Pool()
         data = zip(range(dim), itertools.repeat((band, pca,
-                                                 bounds, ks_gaussian)))
+                                                 bounds, ks_gaussian,
+                                                 seed, use_brute)))
```
Comment (josef-pkt) [inline statsmodels/graphics/functional.py]: Doesn't this use the same seed in each call/repeat?
causing perfect correlation across `dim`

(I never tried to understand the details here.)

Comment (josef-pkt) [reply statsmodels/graphics/functional.py]: seed is just used for differential evolution, so maybe/likely it doesn't matter.

Comment (bashtage) [reply statsmodels/graphics/functional.py]: No, it ultimately calls

https://github.com/statsmodels/statsmodels/blob/c5bf4a6338bea8cdc37b931101cbe133f30543d9/statsmodels/graphics/functional.py#L94

which takes idx as a parameter (which comes from range).  

I'm not trying to fix anything other than to allow it to be deterministic.

Comment (josef-pkt) [reply statsmodels/graphics/functional.py]: But AFAICS it changes the differential evolution behavior.
Before it was using the global random state which continuous the random chain with each call.
Now it uses the same seed in each call to differential_evolution.

I don't think anymore that it matters in this case, but in general using identical seed would be incorrect, e.g. in bootstrap or parallel randomized optimization for the same function. The latter is the case that I thought about initially.

**Comment by josef-pkt** [comment]
I changed the title and added labels to indicate it includes now also a REF/ENH in graphics.

**Comment by bashtage** [comment]
If seed=None then the global is still used everywhere.

**Comment by josef-pkt** [comment]
> If seed=None then the global is still used everywhere.

Good, thanks for pointing it out. I skimmed the docstrings too fast, and didn't think about this.

**Comment by josef-pkt** [comment]
ETA for appveyor looks like 12 hours.

merge when green

**Comment by jbrockmendel** [comment]
@josef-pkt defacto green

**Comment by josef-pkt** [comment]
as green as it gets, merging

@bashtage Thank you

**Pull request merged**